h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstGroup.java

```diff
@@ -751,17 +801,42 @@ public G(int ncols, AGG[] aggs, boolean hasMedian) {
         }
       }
     }
+    
+    private int colsCount() { return _gs.length + _gsStr.length; }
 
     public G fill(int row, Chunk chks[]) {
-      for (int c = 0; c < chks.length; c++) // For all selection cols
-        _gs[c] = chks[c].atd(row); // Load into working array
+      for (int c = 0; c < chks.length; c++) { // For all selection cols
+        Vec vec = chks[c].vec();
```

Comment:
A lot of code in this PR handles switching between the 2 arrays (_gs nad _gsStr). This is due to the fact that we can have an arbitrary order of string/numerical columns on the input. However, in the implementation itself, we can re-order the columns to make sure we will have the exact order we want, eg. numerical first then string ones. Would that be something that would help with making the code less complex?

Reply:
There is nothing that comes to my mind that could solve this problem. Switching between `_gs` and `_gsStr` is done in 2 places - filling the output frame and sorting it. In both cases it is expected to preserve order of the columns. There is a possibility (at least for the "filling" part) of replacing `gbColsTypes`  array with 2 arrays defining order for each of the columns in `_gs` and `_gsStr`.  But I don't know whether it significantly reduces the complexity, if at all.

Reply:
From my point of view, two arrays are ok in this case. One array with both types of values has the same complexity; there should also be `if ` where we will have to decide what to do with string and numeric separately. Maybe some class can solve this, however, I think it is overengineering in this case.